### PR TITLE
NoUnusedVariables validation rule

### PR DESCRIPTION
**Pull request by crisu83**

### CODE REVIEW

File: src/Validation/Rule/NoUnusedVariablesRule.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * No unused variables
+ *
+ * A GraphQL operation is only valid if all variables defined by an operation
+ * are used, either directly or within a spread fragment.
+ */
+class NoUnusedVariablesRule extends AbstractRule
+{
+    /**
+     * @var array|VariableDefinitionNode[]
```
Comment (Jalle19) [inline src/Validation/Rule/NoUnusedVariablesRule.php]: `array` is superflous here

File: src/Validation/Rule/NoUnusedVariablesRule.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    protected $variableDefinitions;
+
+    /**
+     * @inheritdoc
+     */
+    public function enterNode(NodeInterface $node): ?NodeInterface
+    {
+        if ($node instanceof OperationDefinitionNode) {
+            $this->variableDefinitions = [];
```
Comment (Jalle19) [inline src/Validation/Rule/NoUnusedVariablesRule.php]: Is this correct? What if there are more than one operation definition nodes?

File: src/Validation/Rule/NoUnusedVariablesRule.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    public function leaveNode(NodeInterface $node): ?NodeInterface
+    {
+        if ($node instanceof OperationDefinitionNode) {
+            $variableNamesUsed = [];
+            $usages            = $this->validationContext->getRecursiveVariableUsages($node);
+            $operationNameNode = $node->getName();
+            $operationName     = null !== $operationNameNode ? $operationNameNode->getValue() : null;
+
+            /** @var VariableNode $variableNode */
+            foreach ($usages as ['node' => $variableNode]) {
```
Comment (Jalle19) [inline src/Validation/Rule/NoUnusedVariablesRule.php]: This syntax looks weird, are you sure this is correct?

File: src/Validation/Rule/NoUnusedVariablesRule.php (hunk 2 above)
Comment (crisu83) [reply src/Validation/Rule/NoUnusedVariablesRule.php]: Yes, the variables are checked for each OperationDefinitionNode separately.

File: src/Validation/Rule/NoUnusedVariablesRule.php (hunk 3 above)
Comment (crisu83) [reply src/Validation/Rule/NoUnusedVariablesRule.php]: This is correct, it's the destructor syntax.

**Pull request merged**